File: 64-minimum-path-sum/64-minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        
        int m = grid.size();
        int n = grid[0].size();
        
        vector<vector<int>>dp(m, vector<int>(n,-1));
        
        return solve(grid,0,0,m,n,dp);
        
        
    }
    
    int solve(vector<vector<int>>& grid,int i, int j, int m , int n,vector<vector<int>>&dp){
        if(i>=m || j>=n){
            return INT_MAX;
        }
        
        if(i==m-1 && j ==n-1){
            return grid[i][j];
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        
        return dp[i][j]= grid[i][j]+ min(solve(grid,i+1,j,m,n,dp), solve(grid,i,j+1,m,n,dp));
        // return ans;
        
    }
};

File: 64-minimum-path-sum/64-minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
      return backtrack(grid, 0 , 0);
    }
    
     int backtrack(int[][] grid, int r, int c){
        if(r>= grid.length || c>=grid[0].length) return Integer.MAX_VALUE;
        if(r == grid.length - 1 && c == grid[0].length -1) return grid[r][c];
        if(grid[r][c] < 0) return -1*grid[r][c];
        int temp =  grid[r][c] + Math.min(backtrack(grid, r+1, c) ,  backtrack(grid, r, c+1));
        grid[r][c] = -1*temp;
        return temp;
         
     }
    
}